### ISSUE TITLE
Storing information on whether a child object is referenced by id or included recursively

### ISSUE THREAD

**Issue by cwebber** [w3c-dvcg/ld-signatures#7]
(I'm not confident this belongs in an issue, I'm happy to move it somewhere else if someone suggests a better venue.)

tl;dr: Signatures are different depending on whether objects are recursively included or not; servers must keep track of this information to preserve signatures of referenced objects, but how to do this may not have an obviously presented solution, or users may not be aware.

Some users of ActivityPub are exploring the adoption of Linked Data Signatures to ensure the integrity of objects passed around the federated network.  Of course, people don't store these exactly as they received them always.  So imagine if Alyssa P. Hacker offers that Ben Bitdiddle attend her thesis completion party:

``` python
{"@context": ["https://w3id.org/identity/v1", "https://www.w3.org/ns/activitystreams"],
 "id": "https://chatty.example/ben/p/618998",
 "actor": "https://chatty.example/ben/",
 "type": "Accept",
 "to": ["https://social.example/alyssa/", "https://social.example/eva/"],
 "object": {
   "id": "https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473",
   "type": "Invite",
   "actor": "https://social.example/alyssa/",
   "to": ["https://chatty.example/ben/", "https://social.example/eva/"],
   "object": {
     "id": "https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473#object",
     "content": "Come to my thesis completion party!"},
   "target": ["https://chatty.example/ben/", "https://social.example/eva/"],
   "signature": ...},
 "signature": ...}
```

Alyssa and Ben's friend Eva Lu Ator had already received and accepted the invitation, and is happy to see that Ben will also be joining.  She'd like to `Like` the object.  In theory, this activity might look like the following:

``` python
{"@context": ["https://w3id.org/identity/v1", "https://www.w3.org/ns/activitystreams"],
 "id": "https://social.example/alyssa/posts/28ce2e82-10fe-4564-80d4-2fe4d7b0a77d",
 "type": "Like",
 "actor": "https://social.example/eva/",
 "to": ["https://social.example/alyssa/", "https://chatty.example/ben/"],
 "object": {
   "id": "https://chatty.example/ben/p/618998",
   "actor": "https://chatty.example/ben/",
   "type": "Accept",
   "to": ["https://social.example/alyssa/", "https://social.example/eva/"],
   "object": {
     "id": "https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473",
     "type": "Invite",
     "actor": "https://social.example/alyssa/",
     "to": ["https://chatty.example/ben/", "https://social.example/eva/"],
     "object": {
       "id": "https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473#object",
       "content": "Come to my thesis completion party!"},
     "target": ["https://chatty.example/ben/", "https://social.example/eva/"],
     "signature": ...},
   "signature": ...},
 "signature": ...}
```

All well and good.  There's only one problem: Eva's server stores each object independently, not recursively, so it stored Ben's Accept as:

``` python
{"@context": ["https://w3id.org/identity/v1", "https://www.w3.org/ns/activitystreams"],
 "id": "https://chatty.example/ben/p/618998",
 "actor": "https://chatty.example/ben/",
 "type": "Accept",
 "to": ["https://social.example/alyssa/", "https://social.example/eva/"],
 "object": "https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473",
 "signature": ...}
```

In fact, the social protocol did not specify whether objects must be expanded recursively or simply reference other objects.  (It would also have been legitimate to have expanded all of the actors, but this was not done in delivery in this case.)  Eva's database also did not keep track of which fields were expanded recursively or referenced by id.  When Eva constructs her Like object, the server actually constructs it as:

``` python
{"@context": ["https://w3id.org/identity/v1", "https://www.w3.org/ns/activitystreams"],
 "id": "https://social.example/alyssa/posts/28ce2e82-10fe-4564-80d4-2fe4d7b0a77d",
 "type": "Like",
 "actor": "https://social.example/eva/",
 "to": ["https://social.example/alyssa/", "https://chatty.example/ben/"],
 "object": {
   "id": "https://chatty.example/ben/p/618998",
   "actor": "https://chatty.example/ben/",
   "type": "Accept",
   "to": ["https://social.example/alyssa/", "https://social.example/eva/"],
   "object": "https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473",
   "signature": ...},
 "signature": ...}
```

Unfortunately, Eva's reconstruction of Ben's note fails to preserve its signature.  To see why, observe that the normalized data is different depending on whether the object is recursively included:

```
<https://chatty.example/ben/p/618998> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://www.w3.org/ns/activitystreams#Accept> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#actor> <https://chatty.example/ben/> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#object> <https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#to> <https://chatty.example/ben/> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#to> <https://social.example/eva/> .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473#object> <https://www.w3.org/ns/activitystreams#content> "Come to my thesis completion party!" .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://www.w3.org/ns/activitystreams#Invite> .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> <https://www.w3.org/ns/activitystreams#actor> <https://social.example/alyssa/> .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> <https://www.w3.org/ns/activitystreams#object> <https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473#object> .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> <https://www.w3.org/ns/activitystreams#target> <https://chatty.example/ben/> .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> <https://www.w3.org/ns/activitystreams#target> <https://social.example/eva/> .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> <https://www.w3.org/ns/activitystreams#to> <https://chatty.example/ben/> .
<https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> <https://www.w3.org/ns/activitystreams#to> <https://social.example/eva/> .
```

vs

```
<https://chatty.example/ben/p/618998> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://www.w3.org/ns/activitystreams#Accept> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#actor> <https://chatty.example/ben/> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#object> <https://social.example/alyssa/posts/03f26c20-b25a-4086-a9e3-9f841e461473> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#to> <https://chatty.example/ben/> .
<https://chatty.example/ben/p/618998> <https://www.w3.org/ns/activitystreams#to> <https://social.example/eva/> .
```

Missing information means mismatched signature!

What's the right thing to do?  Store every incoming object recursively?  (This seems expensive.)  Specify how much information should be included recursively?  (Seems tedious, and like it shouldn't have to be done.)  Store alongside the stored object which fields are stored recursively and which isn't?  (Could be done, but seems like extra work.)

**Comment by cwebber** [comment]
One way to encode this would be as follows:

```
{"object": {"object": {"object": {}}}}
```

Which would say that in the `Like`, its `object`, as well as that object's `object`, and the `object` of that, should all be recursively embedded, but none of the `actor` fields are, for instance, nor are the `to` fields.  This encoding could be error prone for values with arrays however, where each member of an array may need to be encoded as to whether it is embedded individually.

**Comment by msporny** [comment]
> It seems like something that users will bump into independently though, and perhaps we should have a way to advise them.

Hey @cwebber, thanks for logging the issue. You're right, we should track this somewhere. I'm a bit pressed for time, so this won't be a complete answer to your question (and I don't think there IS a complete answer to this question). That said, we should provide some guidance for folks that need to store signed data in databases.

So, here are the options we employ:

1. Store the entire signed object in the database.
2. Create an application-specific way of serializing to and getting data from the database.
3. Use hash-based identifiers to specify external objects.

## Store the entire signed object in the database

This is the easiest approach because you're guaranteed to not screw up the signature. Store the entire object when you get it, respond with the entire object when it's requested. The signature is always guaranteed to be intact.

## Application-specific database serialization

This approach makes the application determine where to break apart objects. For example, if you have a subobject that belongs in a different database collection/table, you break that out and the application places an indicator in the metadata for the record stating that the object needs to be reconstructing using a value in a separate collection/table/database. This approach can be generalized so you have an application-wide way of deconstructing and reconstructing signed objects.

## Hash-based Deconstruction/Reconstruction

One of the dangers of signing just a URI in Linked Data Signatures is what happens when that URI changes. This can be a benefit when you don't want to sign the data at the URI. However, it can also be a drawback when you DO want to sign the data at the URI. If signing the data at the URI is important, you can swap out the URI for a hash-based identifier OR you can include a hash-based identifier along with the URI. For example, if you had this

```
{
  id: 'https://example.org/people/msporny',
  knows: 'https://example.org/people/cwebber',
  signature: { ... }
}
```

you could store just that, but then the `.../cwebber` object might change. So you may want to do this at the application-database layer:

```
const record = {
  id: 'https://example.org/people/msporny',
  knows: 'ni:///sha-256;s5wBYpPsZ142BXP7fUS/jBL90mxlc78ekcykQVPc8tw=',
  signature: { ... }
}

const recordmeta = {
  reconstruct: ['knows']
}
```

The above would tell the application to reconstruct the 'knows' field based on the hash of the object (which you could get via RDF Dataset Graph Normalization... using something like the jsonld.normalize() call).

You could also do something like this:

```
const record = {
  id: 'https://example.org/people/msporny',
  knows: {
    id: 'https://example.org/people/cwebber',
    subjectHash: 'ni:///sha-256;s5wBYpPsZ142BXP7fUS/jBL90mxlc78ekcykQVPc8tw='
  },
  signature: { ... }
}
```

The above could be written to the database w/o doing deconstruction where you are being very specific about the contents of the `https://example.org/people/cwebber` object, but without having to store it. You're basically saying "as long as you have an object that hashes to `subjectHash`, you can be sure that you have the right data and that data has been digitally signed by whoever signed this database record.

# Other Thoughts

Of these strategies, we tend to usually do number 2 above. Our database code usually knows how to break information apart and store it in the database and then reconstruct it from the database to keep the signatures intact. These sorts of things tend to be very application specific, so we tend to make the application do the heavy lifting of ensuring that the signatures aren't corrupted.

These are the three most helpful design patterns that we've been using over the past 7+ years for signed JSON-LD data. I'm sure there are others, so I'm going to keep this thread open to try and collect other approaches people are discovering.

**Comment by msporny** [comment]
> {"object": {"object": {"object": {}}}}

-1, don't do this; the approach will lead to sadness and heartache.

Keep in mind that any of those objects could change at any time in the database. If you're signing objects AND deconstructing those objects when you store them, you probably need to do content-based addressing of those objects. In general, it's the only way that you can be sure that you're going to grab the right object out of the database. 

The W3C Activity Streams stuff might be a special case in that there are unique IDs for each signed message and the data doesn't change after you send it out there. That said, there are probably exceptions. For example, you could imagine referring to a person object in a message and embeding that person object said message and then deconstructing that person object out into a different database collection/table when storing). If that person object changes over time, your application can get confused as to which version of that person you're referring to... which is why we tend to default to content-based addressing in our applications to be specific about which version of the person object we're referring to.

**Comment by cwebber** [comment]
> The W3C Activity Streams stuff might be a special case in that there are unique IDs for each signed message and the data doesn't change after you send it out there.

It's not in most cases, though it's something [I've thought about](https://dustycloud.org/blog/an-even-more-distributed-activitypub/), but the default in ActivityPub at least (but not necessarily ActivityStreams) is that the social graph is mutable.  Even if we moved to the more git-like pointers-to-specific-objects approach, that's still enough mutability to cause this problem.  So I agree with your assessment!
